### PR TITLE
Replace carder with stats

### PR DESCRIPTION
**Pull request by max-hoffman**
I wanted to separate 1) moving cardinality into stats objects, and 2) changing costing to start at join relations, but I kept finding small bugs and hard a time separating the two. So i did the move, fixed bugs along the way, and then regenerated plans with the changes.

Our current join plans are less useful after we start doing costed join reordering, which makes it kind of annoying that these refactors are dominated by mostly irrelevant plan changes. Still I think the current refactors were necessary, because we need the costing changes and users will still use stats-less tables. Soon I will start adding more unit-esque plan tests that are equally comprehensive but test fewer things, like whether certain plans' cardinalities are accurate, and how closely cost estimates reflect execution runtime. 

Dolt side is passing: https://github.com/dolthub/dolt/pull/7156

### CODE REVIEW

File: sql/memo/memo.go [hunk 1]
```diff
@@ -326,6 +305,20 @@ func (m *Memo) optimizeMemoGroup(grp *ExprGroup) error {
 	}
 	var err error
 	n := grp.First
+	if _, ok := n.(SourceRel); ok {
```
Comment (nicktobey) [inline sql/memo/memo.go:308]: This block needs a comment, and maybe should outlined. I eventually figured out that we want to choose IndexScans over other sourceRel exoressions, but it clutters the rest of the function.

File: sql/memo/rel_props.go [hunk 2]
```diff
@@ -90,6 +91,7 @@ func newRelProps(rel RelExpr) *relProps {
 	default:
 	}
 
+	p.stat = statsForRel(rel)
```
Comment (nicktobey) [inline sql/memo/rel_props.go:94]: Does it matter that the statsForRel call happens before `populateFds` here, and after `populateFds` on line 58?

Should this be outlined?

File: sql/memo/rel_props.go [hunk 3]
```diff
@@ -257,6 +263,99 @@ func (p *relProps) populateFds() {
...
+			return &stats.Statistic{RowCnt: uint64(card)}
+		}
+	case *Max1Row:
+		stat = &stats.Statistic{RowCnt: 1}
+
+	case *EmptyTable:
+		stat = &stats.Statistic{RowCnt: 0}
+
+	case *IndexScan:
+		stat = &stats.Statistic{RowCnt: 10}
```
Comment (nicktobey) [inline sql/memo/rel_props.go:321]: Why do we choose 10 here? A comment explaining the value might be useful.

Comment (nicktobey) [reply sql/memo/rel_props.go:321]: Likewise with the other operations that assume a constant size.

File: sql/memo/rel_props.go [hunk 4]
```diff
@@ -257,6 +263,99 @@ func (p *relProps) populateFds() {
...
+				if card, err := prov.RowCount(rel.Group().m.Ctx, tn.Database().Name(), tn.Name()); err == nil {
+					stat = &stats.Statistic{RowCnt: card}
+					break
+				}
+			}
+		}
+		stat = &stats.Statistic{RowCnt: 1000}
+
+	case *Filter:
+		card := float64(rel.Child.RelProps.GetStats().RowCount()) * .75
```
Comment (nicktobey) [inline sql/memo/rel_props.go:341]: This should probably be outlined into a `filter_selectivity` constant.

**Review by nicktobey — COMMENTED** [review]
Some things that I couldn't leave as comments:

- You should update the docstring for `lookupCandidates` to clarify it returns a TableId instead of a GroupId

**Review by nicktobey — APPROVED** [review]

File: sql/memo/memo.go (hunk 1 above)
Comment (max-hoffman) [reply sql/memo/memo.go:308]: Yeah this points out a key limitation of the way we order the search. Ideally we would generate plan options bottom-up the way we do now, but with the full plan materialized for each node. So we would have a generator capable of keeping track of which plans to check, but the forest of plans and groups should maybe be abstracted from the actual costing logic. The memo representation of inner join would reference indexScans, and a lookupJoin in the same group's LHS can be indexScan/tableScan and RHS be tableScans. The physical operators can share group properties for cardinality/FD tracking, but get rid of the indirection and weird dependencies going on right now.

File: sql/memo/rel_props.go (hunk 2 above)
Comment (max-hoffman) [reply sql/memo/rel_props.go:94]: Pood point, we generate them lazily so as long as the children FDs are populated before this expr it ends up OK. I'll reorder the FDs to be first though.

File: sql/memo/rel_props.go (hunk 4 above)
Comment (max-hoffman) [reply sql/memo/rel_props.go:341]: good idea

**Pull request merged**